## baseline regression models

In [1]:
import sys
import os

sys.path.append(os.path.abspath(".."))

from src.data_loader import load_data
from src.preprocessing import create_features

df = load_data()
df = create_features(df)

df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal,RoomsPerHousehold,BedroomsRatio,PopulationPerHousehold
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526,2.732919,0.146591,126.0
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585,2.956685,0.155797,1138.0
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521,2.957661,0.129516,177.0
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413,2.283154,0.184458,219.0
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422,2.879646,0.172096,259.0


## train test split

In [2]:
from sklearn.model_selection import train_test_split

X = df.drop(
    "MedHouseVal",
    axis=1
)

y = df["MedHouseVal"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

## linear regression baseline

In [3]:
from src.model import get_linear_model
from src.evaluation import evaluate_regression

linear_model = get_linear_model()

linear_model.fit(
    X_train,
    y_train
)

pred_lr = linear_model.predict(
    X_test
)

evaluate_regression(
    y_test,
    pred_lr
)

{'MAE': 0.48743933112258375,
 'RMSE': np.float64(0.6738172773395145),
 'R2': 0.6535205948827303}

## random forest baseline

In [4]:
from src.model import get_random_forest

rf_model = get_random_forest()

rf_model.fit(
    X_train,
    y_train
)

pred_rf = rf_model.predict(
    X_test
)

evaluate_regression(
    y_test,
    pred_rf
)

{'MAE': 0.3279938796632755,
 'RMSE': np.float64(0.5035421491451098),
 'R2': 0.8065072048681708}

## compare baseline models

In [5]:
import pandas as pd

results = pd.DataFrame([
    {
        "Model": "Linear Regression",
        **evaluate_regression(
            y_test,
            pred_lr
        )
    },
    {
        "Model": "Random Forest",
        **evaluate_regression(
            y_test,
            pred_rf
        )
    }
])

results

,Model,MAE,RMSE,R2
0,Linear Regression,0.487439,0.673817,0.653521
1,Random Forest,0.327994,0.503542,0.806507


### Model Comparison Insight

Random Forest significantly outperformed Linear Regression across all evaluation metrics.

The Random Forest model achieved an R² score of 0.807, compared to 0.654 for Linear Regression, indicating that non-linear relationships play an important role in predicting California housing prices.

This result suggests that tree-based models are more suitable for capturing complex interactions between housing features.